### PR TITLE
Update docs and infer_signature

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/13407?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13407/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13407
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
1. Remove the warning message for input example (it is misleading and not needed anymore)
2. Update infer_signature method to accept dataclass directly; ModelSignature constructor shouldn't be exposed to users unless there's no other options.
3. Update docs signature section to include an example with GenAI flavors: dictionary input -> log model -> load back and run predict

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/source/model/signatures.rst [hunk 1]
```diff
@@ -629,6 +629,127 @@ The same signature can be created explicitly as follows:
...
+
+When logging a GenAI flavor model, you can pass the input example to the `input_example` parameter of the `log_model` method.
+MLflow will infer the model signature from the provided input and apply it to the logged model. Additionally, this input example 
+helps validate that the model can run predictions successfully, making it **strongly recommended to always include an input example when logging a GenAI model**.
+
+.. note::
+    By default, fields in the input example are inferred as required. To mark a field as optional, you can provide a list of 
+    input examples where some examples contain None values for the optional fields. This behavior will be improved in future releases.
+
+Take langchain model as an example, you can log a model with signature with below steps:
```
Comment (BenWilson2) [inline docs/source/model/signatures.rst:646]:
```suggestion
Taking a langchain model as an example, you can log a model with signature with 2 simple steps using the models from code feature:
```

File: docs/source/model/signatures.rst [hunk 2]
```diff
@@ -629,6 +629,127 @@ The same signature can be created explicitly as follows:
...
+    model = OpenAI(temperature=0.95)
+
+    chain = (
+        {
+            "question": itemgetter("messages") | RunnableLambda(get_question),
+            "answer": itemgetter("messages") | RunnableLambda(get_answer),
+        }
+        | prompt
+        | model
+        | StrOutputParser()
```
Comment (BenWilson2) [inline docs/source/model/signatures.rst:698]: Can we add a note above explaining just how critical it is to add this parser stage? I think the single most common issue that users have with getting their signatures to work correctly is forgetting or not knowing that they need this step.

Comment (BenWilson2) [inline docs/source/model/signatures.rst:698]: Can we add a note above explaining just how critical it is to add this parser stage? I think the single most common issue that users have with getting their signatures to work correctly is forgetting or not knowing that they need this step.

Comment (BenWilson2) [inline docs/source/model/signatures.rst:698]: Can we add a note above explaining just how critical it is to add this parser stage? I think the single most common issue that users have with getting their signatures to work correctly is forgetting or not knowing that they need this step.

File: docs/source/model/signatures.rst [hunk 3]
```diff
@@ -629,6 +629,127 @@ The same signature can be created explicitly as follows:
...
+            },
+        ]
+    }
+
+    chain_path = "langchain_model.py"
+    with mlflow.start_run():
+        model_info = mlflow.langchain.log_model(
+            lc_model=chain_path, artifact_path="model", input_example=input_example
+        )
+
```
Comment (BenWilson2) [inline docs/source/model/signatures.rst:724]: Can we add an example of defining Optional values? This is another point of confusion (even though it's mentioned in text above, we should show a code snippet for readers who navigate directly to code first)

File: docs/source/model/signatures.rst [hunk 4]
```diff
@@ -629,6 +629,127 @@ The same signature can be created explicitly as follows:
...
+At this stage, the input_example serves multiple purposes: it is used to infer the model signature, 
+validate that the model functions correctly when reloaded using the PyFunc flavor, and to generate 
+a corresponding serving_input_example, which is used to validate the model’s functionality prior to deployment.
```
Comment (BenWilson2) [inline docs/source/model/signatures.rst:727]:
```suggestion
a corresponding serving_input_example. The generated signature is used to validate the model’s functionality prior to deployment as well as providing guidance to callers of the model when it is deployed if they are making requests with invalid data structures.
```

File: docs/source/model/signatures.rst [hunk 5]
```diff
@@ -629,6 +629,127 @@ The same signature can be created explicitly as follows:
...
+3. Load the model back and make predictions:
+
+.. code-block:: python
+
+    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    result = loaded_model.predict(input_example)
+
+    print(result)
+
+**Dataclass object as input is also accepted, use `infer_signature` to get the model signature**:
```
Comment (BenWilson2) [inline docs/source/model/signatures.rst:740]:
```suggestion
**A `Dataclass` object is also a valid way of submitting a signature. Passing a `Dataclass` definition to `infer_signature` will generate the corresponding model signature equivalent**:
```

File: docs/source/model/signatures.rst [hunk 6]
```diff
@@ -629,6 +629,167 @@ The same signature can be created explicitly as follows:
...
+3. Load the model back and make predictions:
+
+.. code-block:: python
+
+    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    result = loaded_model.predict(input_example)
+
+    print(result)
+
+**A Dataclass object is also a valid way of submitting a signature. Passing a Dataclass
```
Comment (harupy) [inline docs/source/model/signatures.rst:779]:
```suggestion
**A Dataclass object is also a valid way of constructing a signature. Passing a Dataclass
```

File: docs/source/model/signatures.rst [hunk 7]
```diff
@@ -629,6 +629,167 @@ The same signature can be created explicitly as follows:
...
+.. code-block:: python
+
+    from mlflow.models import infer_signature
+    from mlflow.models.rag_signatures import ChatCompletionRequest, ChatCompletionResponse
+
+    model_signature = infer_signature(
+        ChatCompletionRequest(),
+        ChatCompletionResponse(),
+    )
+    print(model_signature)
```
Comment (harupy) [inline docs/source/model/signatures.rst:791]: Can we create a custom dataclass and use it?

File: docs/source/model/signatures.rst (hunk 6 above)
Comment (harupy) [reply docs/source/model/signatures.rst:779]: Can we hyperlink `Dataclass` with https://docs.python.org/3/library/dataclasses.html?

File: docs/source/model/signatures.rst [hunk 8]
```diff
@@ -629,6 +629,167 @@ The same signature can be created explicitly as follows:
         inputs=input_schema, outputs=output_schema, params=params_schema
     )
 
+GenAI flavors model signature examples
```
Comment (harupy) [inline docs/source/model/signatures.rst:632]:
```suggestion
Model signature examples for GenAI flavors
```

File: docs/source/model/signatures.rst [hunk 9]
```diff
@@ -629,6 +629,167 @@ The same signature can be created explicitly as follows:
...
+
+.. code-block:: python
+
+    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    result = loaded_model.predict(input_example)
+
+    print(result)
+
+**A Dataclass object is also a valid way of submitting a signature. Passing a Dataclass
+definition to** `infer_signature` **will generate the corresponding model signature equivalent**:
```
Comment (harupy) [inline docs/source/model/signatures.rst:780]:
```suggestion
instance to** `infer_signature` **will generate the corresponding model signature equivalent**:
```

`Dataclass` definition sounds like this:

```
@dataclass
class Foo:
    pass

infer_signature(Foo) # not infer_signature(Foo)
```suggestion
definition to** `infer_signature` **will generate the corresponding model signature equivalent**:
```
```

File: docs/source/model/signatures.rst [hunk 10]
```diff
@@ -629,6 +629,196 @@ The same signature can be created explicitly as follows:
...
+Manually defining the structure is complex and error-prone. Instead, you should rely on the `infer_signature` method to automatically
+generate the model signature based on an input example.
+
+When logging a GenAI flavor model, you can pass the input example to the `input_example` parameter of the `log_model` method.
+MLflow will infer the model signature from the provided input and apply it to the logged model. Additionally, this input example 
+helps validate that the model can run predictions successfully, making it **strongly recommended to always include an input example when logging a GenAI model**.
+
+.. note::
+    By default, fields in the input example are inferred as required. To mark a field as optional, you can provide a list of 
+    input examples where some examples contain None values for the optional fields. This behavior will be improved in future releases.
```
Comment (harupy) [inline docs/source/model/signatures.rst:644]: can you add an example?

File: docs/source/model/signatures.rst [hunk 11]
```diff
@@ -629,6 +629,196 @@ The same signature can be created explicitly as follows:
...
+3. Load the model back and make predictions:
+
+.. code-block:: python
+
+    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    result = loaded_model.predict(input_example)
+
+    print(result)
+
+**A** `Dataclass <https://docs.python.org/3/library/dataclasses.html>`_ **object is also a valid way of submitting a signature. Passing a Dataclass
```
Comment (harupy) [inline docs/source/model/signatures.rst:779]:
```suggestion
**A** `Dataclass <https://docs.python.org/3/library/dataclasses.html>`_ **object is also supported to define a signature. Passing a Dataclass
```

File: docs/source/model/signatures.rst (hunk 10 above)
Comment (serena-ruan) [reply docs/source/model/signatures.rst:644]: There's an example on line 740

Comment (harupy) [reply docs/source/model/signatures.rst:644]: I think that's too far.

Comment (harupy) [reply docs/source/model/signatures.rst:644]:
```suggestion
    By default, fields in the input example are inferred as required. To mark a field as optional, you can provide a list of 
    input examples where some examples contain None values for the optional fields ((e.g. `[{"str": "a"}, {"str": None}]`)). This behavior will be improved in future releases.
```

Comment (harupy) [inline docs/source/model/signatures.rst:644]:
```suggestion
    input examples where some examples contain None values for the optional fields.
```

I don't think we need to mention this.

**Review by harupy — APPROVED** [review]
LGTM

**Pull request merged**